stdpopsim/slim_engine.py

```diff
@@ -948,17 +948,58 @@ def fix_time(event):
     condition_on_allele_frequency = []
     op_id = stdpopsim.ext.ConditionOnAlleleFrequency.op_id
     slim_mutation_ids = []
-    for x in _dfe_to_mtypes(contig).values():
+    slim_mutation_dfe = []
+    for i, x in _dfe_to_mtypes(contig).items():
         slim_mutation_ids.extend([u[0] for u in x])
+        slim_mutation_dfe.extend([i for u in x])
     for ee in extended_events:
-        if hasattr(ee, "mutation_type_id"):
-            mt_id = getattr(ee, "mutation_type_id")
+        if hasattr(ee, "single_site_id"):
+            mut_id = getattr(ee, "single_site_id")
             cls_name = ee.__class__.__name__
-            if mt_id not in slim_mutation_ids:
+            mtype_idx = [
+                j
+                for j, i in enumerate(slim_mutation_dfe)
+                if contig.dfe_list[i].id == mut_id
+            ]
+            n_mtypes = len(mtype_idx)
+            if n_mtypes != 1:
                 raise ValueError(
-                    f"Invalid {cls_name} event with mutation type id "
-                    f"{mt_id} on {contig} with valid IDs {slim_mutation_ids}."
+                    f"The single site mutation type with id '{mut_id}' "
+                    f"referenced by {cls_name} must exist and be uniquely "
+                    f"labelled, but there are {n_mtypes} other mutation types "
+                    f"with this id on {contig}. "
                 )
+            mtype_idx = mtype_idx[0]
+            setattr(ee, "mutation_type_id", slim_mutation_ids[mtype_idx])
+            dfe_idx = slim_mutation_dfe[mtype_idx]
+            dfe_intervals = contig.interval_list[dfe_idx]
+            if dfe_intervals.shape[0] == 0:
+                raise ValueError(
+                    f"The single site id '{mut_id}' referenced by {cls_name} "
+                    f"has no coordinate: it may have been removed by the addition "
+                    f"of an overlapping DFE or a single site with an "
+                    f"identical coordinate."
+                )
+            if (
+                dfe_intervals.shape[0] > 1
+                or dfe_intervals[0, 1] - dfe_intervals[0, 0] != 1
+            ):
+                raise ValueError(
+                    f"The id '{mut_id}' referenced by {cls_name} "
+                    f"refers to a DFE with intervals {dfe_intervals}, "
+                    f"not to a single site."
+                )
+            setattr(ee, "coordinate", dfe_intervals[0, 0])
+            assert len(contig.dfe_list[dfe_idx].mutation_types) == 1
+            mt = contig.dfe_list[dfe_idx].mutation_types[0]
+            if not mt.distribution_type == "f":
+                raise ValueError(
+                    f"The single site id '{mut_id}' referenced by "
+                    f"{cls_name} has a mutation type with fitness "
+                    f"distribution '{mt.distribution_type}', instead "
+                    f"of a fixed fitness coefficient."
+                )
+
```

Comment:
we think that `coordinate` and `mutation_type_id` can be local variables in this loop over extended events, rather than an attribute